activedirectory.sgnl.yaml

```diff
@@ -0,0 +1,229 @@
+# All auth mechanisms specified below must be supported by the specified Adapter.
+auth:
+  - basic:
+      username: "{{Input Required: Username}}"
+      password: "{{Input Required: Password}}"
+entities:
+  User:
+    displayName: User
+    externalId: User
+    description: User Entity in Active Directory
+    pageSize: 999
+    pagesOrderedById: false
+    attributes:
+      - name: dn
+        externalId: dn
+        description: distinguished Name of the user
+        type: String
+        indexed: true
+        uniqueId: true
+      - name: obejctClass
+        externalId: obejctClass
+        type: String
+      - name: sn
+        externalId: sn
+        type: String
+      - name: givenName
+        externalId: givenName
+        type: String
+      - name: distinguishedName
+        externalId: distinguishedName
+        type: String
+      - name: instanceType
+        externalId: instanceType
+        type: Int64
+      - name: whenCreated
+        externalId: whenCreated
+        type: DateTime
+      - name: whenChanged
+        externalId: whenChanged
+        type: DateTime
+      - name: uSNCreated
+        externalId: uSNCreated
+        type: Int64
+      - name: uSNChanged
+        externalId: uSNChanged
+        type: Int64
+      - name: department
+        externalId: department
+        type: String
+      - name: name
+        externalId: name
+        type: String
+      - name: userAccountControl
+        externalId: userAccountControl
+        type: Int64
+      - name: badPwdCount
+        externalId: badPwdCount
+        type: Int64
+      - name: codePage
+        externalId: codePage
+        type: Int64
+      - name: countryCode
+        externalId: countryCode
+        type: Int64
+      - name: badPasswordTime
+        externalId: badPasswordTime
+        type: Int64
+      - name: lastLogoff
+        externalId: lastLogoff
+        type: Int64
+      - name: lastLogon
+        externalId: lastLogon
+        type: Int64
+      - name: pwdLastSet
+        externalId: pwdLastSet
+        type: Int64
+      - name: primaryGroupID
+        externalId: primaryGroupID
+        type: Int64
+      - name: accountExpires
+        externalId: accountExpires
+        type: Int64
+      - name: logonCount
+        externalId: logonCount
+        type: Int64
+      - name: sAMAccountName
+        externalId: sAMAccountName
+        type: String
+      - name: sAMAccountType
+        externalId: sAMAccountType
+        type: Int64
+      - name: userPrincipalName
+        externalId: userPrincipalName
+        type: String
+      - name: objectCategory
+        externalId: objectCategory
+        type: String
+      - name: dSCorePropagationData
+        externalId: dSCorePropagationData
+        type: DateTime
+  Group:
+    description: "Group Entity in Active Directory"
+    displayName: Group
+    externalId: Group
+    pageSize: 999
+    pagesOrderedById: false
+    attributes:
+      - name: dn
```

Comment:
@marc-sgnl Follow-up question:

###  Question: What should be the format of `objectGUID` and `objectSid` ?

Currently, `ldapsearch` returns the base64 encoded value of these IDs wherein
- `objectGUID` is a `UUID`(*16-byte structure*)
- `objectSid` is a `String(Sid)` (*An octet string that contains a security identifier*)

**check ref for more details**

```
# Result from ldapSearch

# marc, Users, corp.sgnl.ai
dn: CN=marc,CN=Users,DC=corp,DC=sgnl,DC=ai
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: user
cn: marc
description: Built-in account for administering the computer/domain
distinguishedName: CN=marc,CN=Users,DC=corp,DC=sgnl,DC=ai
instanceType: 4
whenCreated: 20231013202314.0Z
whenChanged: 20240428061131.0Z
uSNCreated: 8196
memberOf: CN=ADSyncAdmins,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Group Policy Creator Owners,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Domain Admins,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Enterprise Admins,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Schema Admins,CN=Users,DC=corp,DC=sgnl,DC=ai
memberOf: CN=Administrators,CN=Builtin,DC=corp,DC=sgnl,DC=ai
name: marc
objectGUID:: /EV8bsrijE+nd5JP+9Zl9w==
objectSid:: AQUAAAAAAAUVAAAAxpq11ZhA0RW0bN0o9AEAAA==
objectCategory: CN=Person,CN=Schema,CN=Configuration,DC=corp,DC=sgnl,DC=ai
```

### Actual values:
```
objectGUID: fc457c6e-cae2-8c4f-a777-924ffbd665f7
objectSid: S-1-5-21-3585448646-366035096-685599924-500

```


Ref: 
- [objectGUID attribute](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-ada3/937eb5c6-f6b3-4652-a276-5d6bb8979658)
- [Object-Sid attribute](https://learn.microsoft.com/en-us/windows/win32/adschema/a-objectsid)

Reply:
Thanks @Mohitp98  -- We'd expect their Actual Values, this is what AD Admins in particular would be building on top of. The Base64 encoding is a nuance of the ldapsearch tool.

Reply:
@Mohitp98, can you please confirm if the adapter response from the go library we use contains the actual values and not encoded values?

Reply:
Using `ldap-go`, [EntryAttributes](https://pkg.go.dev/github.com/go-ldap/ldap/v3#EntryAttribute) will return value as a [[]string](https://pkg.go.dev/builtin#string) or [[][]byte](https://pkg.go.dev/builtin#byte).

In case of **objectGUID** & **objectSid**, using `[]string` as it is won't be helpful as it is a simple string conversion of actual []byte data. This **does not represent** the actual formatted string.

Example: 

```
objectGUID:"\xfcE|n\xca\xe2\x8cO\xa7w\x92O\xfb\xd6e\xf7"
ojbectSid: "\x01\x05\x00\x00\x00\x00\x00\x05\x15\x00\x00\x00ƚ\xb5\u0558@\xd1\x15\xb4l\xdd(\xf4\x01\x00\x00"
```

To retrieve the actual values of these attributes, we have to handle this using appropriate method:
- ObjectGUID: we can use [ParseBytes](https://pkg.go.dev/github.com/google/uuid#ParseBytes) from github.com/google/uuid
- ObjectSid: We can use [Decode](https://pkg.go.dev/github.com/bwmarrin/go-objectsid@v0.0.0-20191126144531-5fee401a2f37#Decode) from github.com/bwmarrin/go-objectsid